### DISCUSSION TITLE
Detect when a port is dis/connected to a link

### DISCUSSION THREAD

**Discussion by alexandernst** [clientIO/joint#2703] [Q&A]
### Introduction

Hi Roman! Is there a way I can detect if a port has been dis/connected from a link?

Currently I'm using the following 3 methods (workarounds): 

1. In order to detect **when** a link has been **connected** to a port, I subscribe to the`link:connect` event on the paper.
2. In order to detect **when** a link has been **disconnected** from a port, I subscribe to the `remove` event on the graph (since the `link:disconnected` event on the paper won't trigger if the link is removed. btw, is that a bug or a feature?)
3. In order to detect the **current** connected ports (think about the "I'm loading a graph with cells" case), I use a custom initializer that let's me check all the ports of each cell.

With these 3 workarounds combined I'm able to modify the size / color of the ports when a link it dis/connected to/from them. But this is not ideal at all. It feels extremely hacky and it doesn't play nice with the command manager (eg, I have two cells with their ports color set to `red` and radius `4`, I connect them with a link, which triggers some code that sets the port color to `green` and radius `6`, then I undo the operation with the command manager).

I could keep hacking and subscribe to the command manager push / pop events and trigger my custom from there, but that just feels like fixing the wrong problem. Is there a way I can manage port colors / size more easily?

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by alexandernst** [comment]
bump

**Reply by kumilingus** [reply]
Hi @alexandernst, an example of this is on my TODO list. Please bear with me (hopefully this week).

**Reply by alexandernst** [reply]
Hi @kumilingus !! Sure, no problem :) Thank you!! 🙏

**Comment by kumilingus** [comment]
When adjusting the appearance of your shapes based on specific rules, it's a good practice not to include these calculated changes in the _CommandManager_ history. Instead, recalculate them whenever the _CommandManager_ alters the graph structure, such as during an undo or redo operation.

You can implement this approach in two ways:

- Exclude the computed changes from the _CommandManager_ (refer to the [documentation](https://docs.jointjs.com/learn/features/undo-redo#filtering-commands)).
- Avoid applying the computed changes directly to the graph; instead, use highlighters (refer to this [example](https://codesandbox.io/p/sandbox/jj4-raster-bug-forked-2m35jw)).

Personally, I prefer using _highlighters_ to display temporary effects on elements, as it allows me to separate shape definitions from visual effects. This separation is useful when you have different views of the same graph where these effects might not be necessary. For instance, imagine an edit mode where no monitoring is displayed and a view mode with live status updates.

A relevant example can be seen in our [DWDM demo](https://www.jointjs.com/demos/dwdm-circuit).
